File: app/src/main/java/com/gosiewski/taxiappandroid/utils/AuthorizedCallAdapterFactory.java
package com.gosiewski.taxiappandroid.utils;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.Set;

import retrofit2.CallAdapter;
import retrofit2.Retrofit;

public class AuthorizedCallAdapterFactory extends CallAdapter.Factory {
    private final Set<Integer> authorizedRequestsRepository;

    public AuthorizedCallAdapterFactory(Set<Integer> authorizedRequestsRepository) {
        this.authorizedRequestsRepository = authorizedRequestsRepository;
    }

    @Override
    public CallAdapter get(Type returnType, Annotation[] annotations, Retrofit retrofit) {
        Authorized authorizedAnnotation = null;
        for (Annotation annotation : annotations)
            if (Authorized.class == annotation.annotationType()) {
                authorizedAnnotation = (Authorized) annotation;
                break;
            }

        for (CallAdapter.Factory callAdapterFactory : retrofit.callAdapterFactories()) {
            if (callAdapterFactory.getClass() == getClass())
                continue;

            CallAdapter callAdapter = callAdapterFactory.get(returnType, annotations, retrofit);
            if (callAdapter != null)
                if (authorizedAnnotation != null)
                    return new AuthorizedCallAdapter<>(callAdapter, authorizedRequestsRepository);
                else
                    return callAdapter;
        }

        return null;
    }
}

File: app/src/main/java/com/gosiewski/taxiappandroid/utils/AuthorizedCallAdapterFactory.kt
package com.gosiewski.taxiappandroid.utils

import retrofit2.CallAdapter
import retrofit2.Retrofit
import java.lang.reflect.Type


class AuthorizedCallAdapterFactory(private val authorizedRequestsRepository: MutableSet<Int>) : CallAdapter.Factory() {

    override fun get(returnType: Type, annotations: Array<Annotation>, retrofit: Retrofit): CallAdapter<*, *>? {
        val authorizedAnnotation: Authorized? = annotations
                .firstOrNull { Authorized::class.java == it.annotationClass }
                ?.let { it as Authorized }

        retrofit.callAdapterFactories()
                .filter { it.javaClass != javaClass }
                .mapNotNull { it.get(returnType, annotations, retrofit) }
                .forEach {
                    return if (authorizedAnnotation != null)
                        AuthorizedCallAdapter(it, authorizedRequestsRepository)
                    else
                        it
                }

        return null
    }
}